File: app/src/main/java/it/liceoarzignano/bold/ui/recyclerview/RecyclerViewExt.java
package it.liceoarzignano.bold.ui.recyclerview;

import android.content.Context;
import android.content.res.TypedArray;
import android.support.v7.widget.DefaultItemAnimator;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.AttributeSet;

import it.liceoarzignano.bold.R;
import it.liceoarzignano.bold.utils.UiUtils;

public class RecyclerViewExt extends RecyclerView {
    private OnItemTouchListener mItemTouchListener = null;

    public RecyclerViewExt(Context context) {
        super(context);
        setup(context, null);
    }

    public RecyclerViewExt(Context context, AttributeSet attrs) {
        super(context, attrs);
        setup(context, attrs);
    }

    public RecyclerViewExt(Context context, AttributeSet attrs, int style) {
        super(context, attrs, style);
        setup(context, attrs);
    }

    @Override
    public void addOnItemTouchListener(OnItemTouchListener listener) {
        if (mItemTouchListener != null) {
            removeOnItemTouchListener(mItemTouchListener);

        }
        mItemTouchListener = listener;

        super.addOnItemTouchListener(listener);
    }

    private void setup(Context context, AttributeSet attrs) {
        setItemAnimator(new DefaultItemAnimator());
        if (attrs == null) {
            return;
        }

        TypedArray array = context.obtainStyledAttributes(attrs,
                R.styleable.RecyclerViewExt, 0, 0);

        try {
            if (array.getBoolean(R.styleable.RecyclerViewExt_horizontalMode, false)) {
                setLayoutManager(new LinearLayoutManager(context, HORIZONTAL, false));
            } else if (UiUtils.isPhone(context)) {
                setLayoutManager(new LinearLayoutManager(context));
            } else {
                setLayoutManager(array.getBoolean(R.styleable.RecyclerViewExt_tabletUI, false) ?
                        new GridLayoutManager(context, 2) : new LinearLayoutManager(context));
            }
        } finally {
            array.recycle();
        }
    }
}

File: app/src/main/kotlin/it/liceoarzignano/bold/ui/recyclerview/RecyclerViewExt.kt
package it.liceoarzignano.bold.ui.recyclerview

import android.content.Context
import android.util.AttributeSet
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import it.liceoarzignano.bold.R
import it.liceoarzignano.bold.utils.UiUtils

class RecyclerViewExt : RecyclerView {
    private lateinit var mItemTouchListener: RecyclerView.OnItemTouchListener

    constructor(context: Context) : super(context) {
        setup(context, null)
    }

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
        setup(context, attrs)
    }

    constructor(context: Context, attrs: AttributeSet, style: Int) : super(context, attrs, style) {
        setup(context, attrs)
    }

    override fun addOnItemTouchListener(listener: androidx.recyclerview.widget.RecyclerView.OnItemTouchListener) {
        if (::mItemTouchListener.isInitialized) {
            removeOnItemTouchListener(mItemTouchListener)
        }
        mItemTouchListener = listener

        super.addOnItemTouchListener(listener)
    }

    private fun setup(context: Context, attrs: AttributeSet?) {
        itemAnimator = androidx.recyclerview.widget.DefaultItemAnimator()
        if (attrs == null) {
            return
        }

        val array = context.obtainStyledAttributes(attrs,
                R.styleable.RecyclerViewExt, 0, 0)

        layoutManager = try {
            when {
                array.getBoolean(R.styleable.RecyclerViewExt_horizontalMode, false) ->
                    LinearLayoutManager(context, HORIZONTAL, false)
                UiUtils.isPhone(context) ->
                    LinearLayoutManager(context)
                else ->
                    if (array.getBoolean(R.styleable.RecyclerViewExt_tabletUI, false))
                        GridLayoutManager(context, 2)
                    else
                        LinearLayoutManager(context)
            }
        } finally {
            array.recycle()
        }
    }
}
